### PR TITLE
GH-36975: [C++][FlightRPC] Skip unknown fields, don't crash

### LINKED ISSUE #36975: [C++][FlightRPC] FlightDataDeserialize incorrectly rejects unknown field tags
### Describe the bug, including details regarding any error messages, version, and platform.

https://github.com/apache/arrow/issues/32276#issuecomment-1660143746

https://github.com/apache/arrow/blob/66a6136c02ad471585a5c38589ba6d779ad28d16/cpp/src/arrow/flight/transport/grpc/serialization_internal.cc#L379-L380

### Component(s)

C++, FlightRPC

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

We should skip unknown fields instead of crashing, for forwards compatibility. 

### What changes are included in this PR?

Skip unknown fields in the FlightData deserializer.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* Closes: #36975

### CODE REVIEW

File: cpp/src/arrow/flight/flight_internals_test.cc [hunk 1]
```diff
@@ -651,6 +654,38 @@ TEST_F(TestCookieParsing, CookieCache) {
...
+
+  auto serialized = raw.SerializeAsString();
+
+  grpc_slice slice = grpc_slice_from_copied_buffer(serialized.data(), serialized.size());
+  // gRPC requires that grpc_slice and grpc::Slice have the same representation
+  grpc::ByteBuffer buffer(reinterpret_cast<const grpc::Slice*>(&slice), /*nslices=*/1);
+
+  flight::internal::FlightData out;
+  auto status = flight::transport::grpc::FlightDataDeserialize(&buffer, &out);
+  ASSERT_TRUE(status.ok());
```
Comment (kou) [inline cpp/src/arrow/flight/flight_internals_test.cc:680]: Can we use `ASSERT_OK()` here?
```suggestion
  ASSERT_OK(flight::transport::grpc::FlightDataDeserialize(&buffer, &out));
```

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/flight/flight_internals_test.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/flight_internals_test.cc:680]: Ah, this is a gRPC status, not an Arrow status, so it won't compile

Comment (kou) [reply cpp/src/arrow/flight/flight_internals_test.cc:680]: Ah, sorry!

**Pull request merged**